### PR TITLE
feat: Rewrite simp_mem to build a new expression, thereby localizing the effects of rewrites [6/?]

### PR DESCRIPTION
**Pull request by bollu**
### Issues:

Resolves 

Remove this section if this PR doesn't address any open issues.

### Description:

A detailed description of your contribution. Why is this change
necessary?

### Testing:

What tests have been run? Did `make all` succeed for your changes? Was
conformance testing successful on an Aarch64 machine?

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Arm/Memory/SeparateAutomation.lean [hunk 1]
```diff
@@ -277,7 +315,14 @@ partial def SimpMemM.simplifyGoal (g : MVarId) (hyps : Array Memory.Hypothesis)
   SimpMemM.withContext g do
     let gt ← g.getType
     withTraceNode m!"Simplifying goal." do
-        SimpMemM.simplifyExpr (← whnf gt) hyps
+        -- TODO: I changed this to WHNF!
+        -- let out ← SimpMemM.simplifyExpr (← whnf gt) hyps
+        let some out ← SimpMemM.simplifyExpr gt hyps
+          | return ()
+        check out.eNew
+        check out.eqProof
```
Comment (shigoel) [inline Arm/Memory/SeparateAutomation.lean:323]: Note that if this is a bottleneck performance wise, we can turn it off.

File: Arm/Memory/SeparateAutomation.lean [hunk 2]
```diff
@@ -191,81 +191,119 @@ def getConfig : SimpMemM SimpMemConfig := do
...
+    | (some fResult, none) =>
+        let outResult ← mkCongrFun fResult.eqProof x
+        return some ⟨e.updateApp! fResult.eNew x, outResult⟩
+    | (some fResult, some xResult) =>
+        let outResult ← mkCongr fResult.eqProof xResult.eqProof
+        return some ⟨e.updateApp! fResult.eNew xResult.eNew, outResult⟩
+    | _ => return none
+    -- let outResult ← mkCongr fResult.eqProof xResult.eqProof
+    -- -- | I think I see where the problem is. here, I should have updated with the other result.
+    -- return ⟨e.updateApp! f x, outResult⟩
```
Comment (shigoel) [inline Arm/Memory/SeparateAutomation.lean:305]: Trash this?

File: Arm/Memory/SeparateAutomation.lean [hunk 3]
```diff
@@ -191,81 +191,119 @@ def getConfig : SimpMemM SimpMemConfig := do
...
+  for hyp in hyps do
+    let Hypothesis.read_eq hReadEq := hyp
+      | continue
+    let some subsetProof ← SimpMemM.findOverlappingReadHypAux hyps er hReadEq
+      | continue
+    return some ⟨hReadEq, subsetProof⟩
+  return none
+
+
+def Expr.kindStr (e : Expr) : String :=
```
Comment (shigoel) [inline Arm/Memory/SeparateAutomation.lean:215]: Is this different from `Lean.Expr.ctorName`?

**Comment by shigoel** [comment]
@bollu Could you turn off all the tracing options so that the build is less noisy?

File: Arm/Memory/SeparateAutomation.lean (hunk 3 above)
Comment (bollu) [reply Arm/Memory/SeparateAutomation.lean:215]: Very nice, I didn't notice this API. Switching to it now!

**Review by shigoel — APPROVED** [review]

**Pull request merged**